comment/views/blocker.py

```diff
@@ -0,0 +1,58 @@
+from django.views import View
+
+from comment.models import BlockedUser, BlockedUserHistory, Comment
+from comment.mixins import CanBlockUsersMixin
+from comment.responses import UTF8JsonResponse, DABResponseData
+from comment.messages import BlockUserMSG
+
+
+class BaseToggleBlockingView(DABResponseData):
+    response_class = None
+
+    def get_response_class(self):
+        assert self.response_class is not None, (
+                "'%s' should either include a `response_class` attribute, "
+                "or override the `get_response_class()` method."
+                % self.__class__.__name__
+        )
+        return self.response_class
+
+    def post(self, request, *args, **kwargs):
+        response_class = self.get_response_class()
+        request_data = request.POST or getattr(request, 'data', {})
+        comment_id = request_data.get('comment_id', None)
+        try:
+            comment = Comment.objects.get(id=int(comment_id))
+        except (Comment.DoesNotExist, ValueError, TypeError):
+            self.error = {
+                'detail': BlockUserMSG.INVALID
+            }
+            self.status = 400
+            return response_class(self.json(), status=self.status)
+
+        blocked_user, created = BlockedUser.objects.get_or_create_blocked_user_for_comment(comment)
+
+        if not created:
+            blocked_user.blocked = not blocked_user.blocked
+        blocked_user.save()
+
+        reason = request_data.get('reason', None)
+        if blocked_user.blocked and not reason:
+            reason = comment.content
+
+        BlockedUserHistory.objects.create(
+            blocked_user=blocked_user,
+            blocker=request.user,
+            reason=reason,
+            state=int(blocked_user.blocked)
+        )
```

Comment:
This could probably be a part of the `BlockedUserHistory` manager.
```python
def create_history(*, user, comment, blocker, reason=None):
    if user.is_blocked and not reason:
         reason = comment.content
    self.create(blocked_user=user, blocker=blocker, reason=reason)
```
It can be called appropriately from the view.

Also, the setting of `state`  should ideally be inside the model class method:
```python
# models/blocker.py

def _set_state(self):
    self.state = int(self.blocked_user.blocked)

def save(self, *args, **kwargs):
    self._set_state()
    super().save(*args, **kwargs)
```

Reply:
Agree for the manager part.

I am not sure why I need `_set_state` at all since the state can be set on the record creation only. There is no case to change an entry state at any point!